File: sdks/java/src/main/java/com/cavs/sdk/CavsOptions.java
package com.cavs.sdk;

import com.cavs.sdk.nativebridge.FfmNativeBridge;
import com.cavs.sdk.nativebridge.NativeBridge;

/** Construction options for a {@link CavsClient}. */
public final class CavsOptions {

    private final NativeBridge bridge;

    private CavsOptions(NativeBridge bridge) {
        this.bridge = bridge;
    }

    /** Defaults: the FFM native bridge. */
    public static CavsOptions defaults() {
        return new CavsOptions(new FfmNativeBridge());
    }

    public static Builder builder() {
        return new Builder();
    }

    NativeBridge bridge() {
        return bridge;
    }

    public static final class Builder {
        private NativeBridge bridge;

        /** Override the native backend (e.g. a JNA bridge or a test double). */
        public Builder bridge(NativeBridge b) {
            this.bridge = b;
            return this;
        }

        public CavsOptions build() {
            return new CavsOptions(bridge != null ? bridge : new FfmNativeBridge());
        }
    }
}

File: sdks/kotlin/src/main/kotlin/com/cavs/sdk/CavsOptions.kt
package com.cavs.sdk

import com.cavs.sdk.nativebridge.FfmNativeBridge
import com.cavs.sdk.nativebridge.NativeBridge

/** Construction options for a [CavsClient]. */
class CavsOptions private constructor(internal val bridge: NativeBridge) {

    class Builder {
        private var bridge: NativeBridge? = null

        /** Override the native backend (e.g. a test double). */
        fun bridge(bridge: NativeBridge) = apply { this.bridge = bridge }

        fun build() = CavsOptions(bridge ?: FfmNativeBridge())
    }

    companion object {
        /** Defaults: the FFM native bridge. */
        fun defaults() = CavsOptions(FfmNativeBridge())

        fun builder() = Builder()
    }
}
